File: OpenBurnBarCore/Sources/OpenBurnBarKernel/SharedModels/MobileStreamsListPolicy.swift
import Foundation

/// Distinct Streams list states. Error must never render as empty.
public enum MobileStreamsListPresentation: String, Sendable, Equatable {
    case loading
    case failed
    case empty
    case locked
    case ready
    case paginating
}

public struct MobileStreamsPageOutcome: Sendable, Equatable {
    public let rowCount: Int
    public let hasMore: Bool
    public let canLoadNext: Bool
    public let presentation: MobileStreamsListPresentation

    public init(
        rowCount: Int,
        hasMore: Bool,
        canLoadNext: Bool,
        presentation: MobileStreamsListPresentation
    ) {
        self.rowCount = rowCount
        self.hasMore = hasMore
        self.canLoadNext = canLoadNext
        self.presentation = presentation
    }
}

/// Pagination / empty / error / lock for Streams. Source: iOS `ActivityStore`.
public enum MobileStreamsListPolicy {
    public static func presentation(
        isLoading: Bool,
        failed: Bool,
        isEmpty: Bool,
        entitled: Bool,
        hasMore: Bool,
        isPaginating: Bool,
        searchFailed: Bool
    ) -> MobileStreamsListPresentation {
        if !entitled { return .locked }
        if failed || searchFailed { return .failed }
        if isLoading && isEmpty { return .loading }
        if isEmpty { return .empty }
        if isPaginating && hasMore { return .paginating }
        return .ready
    }

    /// `hasMore` follows iOS: a present cursor means another page exists.
    public static func pageOutcome(
        accumulatedCount: Int,
        pageCount: Int,
        pageSize: Int,
        lastCursorPresent: Bool,
        failed: Bool,
        entitled: Bool = true,
        isLoading: Bool = false
    ) -> MobileStreamsPageOutcome {
        let hasMore = !failed && lastCursorPresent && pageCount == pageSize
        let presentation = presentation(
            isLoading: isLoading,
            failed: failed,
            isEmpty: accumulatedCount == 0,
            entitled: entitled,
            hasMore: hasMore,
            isPaginating: false,
            searchFailed: false
        )
        return MobileStreamsPageOutcome(
            rowCount: accumulatedCount,
            hasMore: hasMore,
            canLoadNext: hasMore && !isLoading && !failed,
            presentation: presentation
        )
    }
}

File: android/app/src/main/java/com/openburnbar/data/policy/MobileStreamsListPolicy.kt
package com.openburnbar.data.policy

/** Distinct Streams list states. Error must never render as empty. */
enum class MobileStreamsListPresentation(val wire: String) {
    LOADING("loading"),
    FAILED("failed"),
    EMPTY("empty"),
    LOCKED("locked"),
    READY("ready"),
    PAGINATING("paginating"),
}

data class MobileStreamsPageOutcome(
    val rowCount: Int,
    val hasMore: Boolean,
    val canLoadNext: Boolean,
    val presentation: MobileStreamsListPresentation,
)

/** Pagination / empty / error / lock for Streams. Source: iOS `ActivityStore`. */
object MobileStreamsListPolicy {
    fun presentation(
        isLoading: Boolean,
        failed: Boolean,
        isEmpty: Boolean,
        entitled: Boolean,
        hasMore: Boolean,
        isPaginating: Boolean,
        searchFailed: Boolean,
    ): MobileStreamsListPresentation = when {
        !entitled -> MobileStreamsListPresentation.LOCKED
        failed || searchFailed -> MobileStreamsListPresentation.FAILED
        isLoading && isEmpty -> MobileStreamsListPresentation.LOADING
        isEmpty -> MobileStreamsListPresentation.EMPTY
        isPaginating && hasMore -> MobileStreamsListPresentation.PAGINATING
        else -> MobileStreamsListPresentation.READY
    }

    fun pageOutcome(
        accumulatedCount: Int,
        pageCount: Int,
        pageSize: Int,
        lastCursorPresent: Boolean,
        failed: Boolean,
        entitled: Boolean = true,
        isLoading: Boolean = false,
    ): MobileStreamsPageOutcome {
        val hasMore = !failed && lastCursorPresent && pageCount == pageSize
        return MobileStreamsPageOutcome(
            rowCount = accumulatedCount,
            hasMore = hasMore,
            canLoadNext = hasMore && !isLoading && !failed,
            presentation = presentation(
                isLoading = isLoading,
                failed = failed,
                isEmpty = accumulatedCount == 0,
                entitled = entitled,
                hasMore = hasMore,
                isPaginating = false,
                searchFailed = false,
            ),
        )
    }
}
